Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -939,26 +977,398 @@ Query Serializer::DecodeQueryTarget(
   }
 
   ResourcePath path = DecodeQueryPath(reader, DecodeString(proto.parent));
-  StructuredQuery query = DecodeStructuredQuery(proto.structured_query);
+  const google_firestore_v1_StructuredQuery& query = proto.structured_query;
 
-  size_t from_count = query.from.size();
+  CollectionGroupId collection_group;
+  size_t from_count = query.from_count;
   if (from_count > 0) {
     if (from_count != 1) {
       reader->Fail(
           "StructuredQuery.from with more than one collection is not "
           "supported.");
+      return Query::Invalid();
+    }
+
+    google_firestore_v1_StructuredQuery_CollectionSelector& from =
+        query.from[0];
+    auto collection_id = DecodeString(from.collection_id);
+    if (from.all_descendants) {
+      collection_group = std::make_shared<const std::string>(collection_id);
+    } else {
+      path = path.Append(collection_id);
+    }
+  }
+
+  FilterList filter_by;
+  if (query.where.which_filter_type != 0) {
+    filter_by = DecodeFilters(reader, query.where);
+  }
+
+  OrderByList order_by;
+  if (query.order_by_count > 0) {
+    order_by = DecodeOrderBys(reader, query.order_by, query.order_by_count);
+  }
+
+  int32_t limit = Query::kNoLimit;
+  if (query.has_limit) {
+    limit = query.limit.value;
+  }
+
+  std::shared_ptr<Bound> start_at;
+  if (query.start_at.values_count > 0) {
+    start_at = DecodeBound(reader, query.start_at);
+  }
+
+  std::shared_ptr<Bound> end_at;
+  if (query.end_at.values_count > 0) {
+    end_at = DecodeBound(reader, query.end_at);
+  }
+
+  return Query(std::move(path), std::move(collection_group),
+               std::move(filter_by), std::move(order_by), limit,
+               std::move(start_at), std::move(end_at));
+}
+
+google_firestore_v1_StructuredQuery_Filter Serializer::EncodeFilters(
+    const FilterList& filters) const {
+  google_firestore_v1_StructuredQuery_Filter result{};
+
+  auto is_field_filter = [](const Filter& f) { return f.IsAFieldFilter(); };
+  auto first = std::find_if(filters.begin(), filters.end(), is_field_filter);
+  size_t filters_count =
```

Comment:
This statement is sufficiently complicated that I had a hard time understanding what you were doing. Why not just:

```c++
size_t filters_count = absl::c_count_if(filters, is_field_filter);
if (filters_count == 1) {
  auto first = absl::c_find_if(filters, is_field_filter);
  // ...
}
```

Reply:
This is more efficient because it avoids traversing the container twice (remember that this rewrite came out of a discussion about performance). But you're right that it's overkill.

Changed to the way you suggested. Ironically, this is what I had originally when I first started work on this function (then I changed it to the Objective-C way to be more "consistent", and so on).

Reply:
For what it's worth, my perspective on this is that it's usually worth the effort to remove extra copies but rarely worth the effort to absolutely minimize the number of reads. The former can frequently be expressed quite readably (as was done here) while the latter can lead to some magic-looking or low-level contortions.

As always, feel free to push back.